### PR TITLE
Fix Decompression Failure on Large Data (mysqldb)

### LINKED ISSUE #3835: [Bug / Regression] mysql_db import fail to decompress dumps
##### ISSUE TYPE

Bug Report (regression)
##### COMPONENT NAME

mysql_db (import)
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /Users/benjamin/xxxxxxx/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

```
[defaults]
host_key_checking = False
forks=500
pipelining=True
retry_files_enabled = False

gathering = smart
fact_caching = jsonfile
fact_caching_connection = ./.tmp/ansible
fact_caching_timeout = 3600

[ssh_connection]
control_path = /tmp/ansible-ssh-%%h-%%p-%%r
ssh_args = -o ControlMaster=auto -o ControlPersist=30m
```
##### OS / ENVIRONMENT

Deployment OS: Mac OS X 10.11.5, python v2.7.11 with pyenv/virtualenv
Destination OS: ubuntu jessie/sid
##### SUMMARY

Using `mysql_db` to import Gzipped or Bzipped SQL dumps used to work like a charm with `ansible 2.0.2.0`
Now, compressed imports fail with a `broken pipe` error, either if `.gz` or .`bz2`
Strangely, this does not happen on a small (compressed) file (1.8k gzip compressed, 6k uncompressed)
Maybe related to https://blog.nelhage.com/2010/02/a-very-subtle-bug/
##### STEPS TO REPRODUCE

Try to import a compressed (large enough) SQL dump with mysql_db.
Failure happen with a 3.5MB gzip compressed / 20MB uncompressed dump.

```
- name: Restore database
  mysql_db:
  args:
    name: my_db
    state: import
    target: /path_to_backups/backup-pre-release.sql.bz2
    login_host: "{{ db.host }}"
    login_port: "{{ db.port }}"
    login_user: "{{ db.user }}"
    login_password: "{{ db.passwd }}"
```
##### EXPECTED RESULTS

Import should just work.
##### ACTUAL RESULTS

```
fatal: [xxxxxx]: FAILED! => {"changed": false, "failed": true, 
"msg": 
"bzip2: I/O or other error, bailing out.  Possible reason follows.
 bzip2: Broken pipe
       Input file = /opt/xxxxxx/backup-pre-release.sql.bz2, output file = (stdout)
"}
```

### PR DESCRIPTION
**Pull request by paolopaolopaolo**
##### ISSUE TYPE

- Bugfix Pull Request
##### COMPONENT NAME

database/mysql/mysqldb
##### ANSIBLE VERSION

```
ansible 2.0.1.0
  config file =(overridden with CONFIGURATIONS BELOW)
  configured module search path = Default w/o overrides
```
##### CONFIGURATIONS

```
[defaults]
host_key_checking = False
forks=500
pipelining=True
retry_files_enabled = False

gathering = smart
fact_caching = jsonfile
fact_caching_connection = ./.tmp/ansible
fact_caching_timeout = 3600

[ssh_connection]
control_path = /tmp/ansible-ssh-%%h-%%p-%%r
ssh_args = -o ControlMaster=auto -o ControlPersist=30m
```
##### SUMMARY

- Reads from unzipping process's stdout and performs another subprocess for each line 
- `subprocess.PIPE` can't transfer extremely large batches of data to other subprocesses. This fix prevents us from using `subprocess.PIPE` irresponsibly.
- Fixes #3835
##### FAILING OUTPUT (as of June 4, 2016)

```
FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "collation": "",
            "config_file": "/root/.my.cnf",
            "connect_timeout": 30,
            "encoding": "",
            "login_host": "127.0.0.1",
            "login_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "login_port": 3306,
            "login_unix_socket": null,
            "login_user": "bob",
            "name": "stuff",
            "quick": true,
            "single_transaction": false,
            "ssl_ca": null,
            "ssl_cert": null,
            "ssl_key": null,
            "state": "import",
            "target": "/server/dataJun-2-2016.sql.bz2"
        },
        "module_name": "mysql_db"
    },
    "msg": "\nbzip2: I/O or other error, bailing out.  Possible reason follows.\nbzip2: Broken pipe\n\tInput file = /server/dataJun-2-2016.sql.bz2, output file = (stdout)\n"
}
```
##### PASSING OUTPUT

```
{"msg": "","invocation": {"module_args": {"login_port": 3306, "ssl_cert": null, "config_file": "~/.my.cnf", "name": "stuff", "encoding": "", "login_user": "bob", "login_host": "127.0.0.1", "ssl_key": null, "ssl_ca": null, "login_unix_socket": null, "state": "import", "login_password": "12345", "collation": "", "target": "/server/dataJun-2-2016.sql.bz2"}}, "changed": true, "db": "stuff"}
```

@nitzmahone

### CODE REVIEW

File: database/mysql/mysql_db.py [hunk 1]
```diff
@@ -198,15 +198,22 @@ def db_import(module, host, user, password, db_name, target, all_databases, port
...
+                    line_buffer = ''
+                    cmd.append('-e')
+                    cmd.append('{}'.format(line))
+                    p2 = subprocess.Popen(cmd, stderr=subprocess.PIPE)
+                    for err_line in p2.stderr:
+                        err_message = '\n'.join((err_message, err_line))
+                    cmd = cmd[:-2]
+                else:
+                    line_buffer = line
+        return 0, '', err_message
```
Comment (nitzmahone) [inline database/mysql/mysql_db.py]: looks like you're ignoring the returncode(s) from p1/p2?

**Comment by paolopaolopaolo** [comment]
@nitzmahone, re-included returncodes! we only evaluate them now once the processes have finished (p2 returncode is the last code, any errors are passed in to stderror)

**Comment by gregdek** [comment]
Thanks @paolopaolopaolo. To the current maintainers, @Jmainguy please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by nitzmahone** [comment]
As we discussed at Pycon- I suspect this solution is a bit Draconian, and may cause perf issues and other unanticipated problems by executing the imports statement-wise instead of in bulk. If this issue is reliably reproducible, delegating the import entirely to the shell (ie, just have the module build the shell command with pipes and exec it, then check the pipe statuses to get returncodes, etc) is probably the right thing to do (getting Python out of the middle as the pipe broker). Anyone else have thoughts?

**Comment by paolopaolopaolo** [comment]
@nitzmahone I was able to reproduce this error today (turned out my test compressed sql file shrank in size at some point so I was getting false positives all day that day). Implemented the changes that you recommended and got the PASSING OUTPUT that I posted in the PR message above.

**Comment by Jmainguy** [comment]
This has been reopened in ansible github, this can be closed out.

**Pull request closed without merging**